// state-pool/GlobalState.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createGlobalstate = exports.GlobalState = void 0;
var immer_1 = __importStar(require("immer"));
var GlobalState = /** @class */ (function () {
    function GlobalState(initialValue) {
        this.value = initialValue;
        this.subscriptions = [];
        this.persist = false;
    }
    GlobalState.prototype.getValue = function (selector) {
        if (selector) {
            return selector(this.value);
        }
        return this.value;
    };
    GlobalState.prototype.refresh = function () {
        this.subscriptions.forEach(function (subscription) {
            if (subscription.reRender) {
                subscription.reRender();
            }
        });
    };
    GlobalState.prototype.setValue = function (newValue, config) {
        if (config === void 0) { config = {}; }
        if (newValue === undefined) {
            this.__updateValue(function (draftVal) { return immer_1.nothing; }, config);
        }
        else if (Object.prototype.toString.call(newValue) === '[object Function]') {
            var reducer = newValue;
            this.setValue(reducer(this.getValue(config.selector)), config);
        }
        else {
            this.__updateValue(function (draftVal) { return newValue; }, config);
        }
    };
    GlobalState.prototype.updateValue = function (updater, config) {
        if (config === void 0) { config = {}; }
        var updaterWrap = function (draftState) {
            // This wrap is for disabling setting returned value
            // We don't allow returned value to be set(just return undefined)
            updater(draftState);
        };
        this.__updateValue(updaterWrap, config);
    };
    GlobalState.prototype.__updateValue = function (updater, config) {
        if (config === void 0) { config = {}; }
        var selector = config.selector;
        var patcher = config.patcher;
        var oldState = this.value;
        var newState;
        if (selector && patcher) {
            var nodeValue_1 = (0, immer_1.default)(selector(oldState), updater);
            newState = (0, immer_1.default)(oldState, function (draftCurrentState) {
                // Avoid setting returns
                patcher(draftCurrentState, nodeValue_1);
            });
        }
        else {
            newState = (0, immer_1.default)(oldState, updater);
        }
        this.value = newState;
        if (newState !== oldState) {
            // There's a new update
            this.subscriptions.forEach(function (subscription) {
                if (subscription.selector(newState) !== subscription.selector(oldState)) {
                    // Node value has changed
                    subscription.observer(subscription.selector(newState));
                }
            });
        }
    };
    GlobalState.prototype.subscribe = function (itemToSubscribe) {
        var _this = this;
        var _itemToSubscribe;
        if (Object.prototype.toString.call(itemToSubscribe) === '[object Function]') {
            _itemToSubscribe = {
                observer: itemToSubscribe,
                selector: function (state) { return state; }
            };
        }
        else {
            _itemToSubscribe = itemToSubscribe;
        }
        if (this.subscriptions.indexOf(_itemToSubscribe) === -1) {
            // Subscribe a component to this global state
            this.subscriptions.push(_itemToSubscribe);
        }
        ;
        var unsubscribe = function () {
            _this.subscriptions = _this.subscriptions.filter(function (subscription) { return (subscription !== _itemToSubscribe); });
        };
        return unsubscribe;
    };
    return GlobalState;
}());
exports.GlobalState = GlobalState;
function createGlobalstate(initialValue) {
    return new GlobalState(initialValue);
}
exports.createGlobalstate = createGlobalstate;
